// roku-deploy/dist/Errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var InvalidDeviceResponseCodeError = /** @class */ (function (_super) {
    __extends(InvalidDeviceResponseCodeError, _super);
    function InvalidDeviceResponseCodeError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, InvalidDeviceResponseCodeError.prototype);
        return _this;
    }
    return InvalidDeviceResponseCodeError;
}(Error));
exports.InvalidDeviceResponseCodeError = InvalidDeviceResponseCodeError;
var UnauthorizedDeviceResponseError = /** @class */ (function (_super) {
    __extends(UnauthorizedDeviceResponseError, _super);
    function UnauthorizedDeviceResponseError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, UnauthorizedDeviceResponseError.prototype);
        return _this;
    }
    return UnauthorizedDeviceResponseError;
}(Error));
exports.UnauthorizedDeviceResponseError = UnauthorizedDeviceResponseError;
var UnparsableDeviceResponseError = /** @class */ (function (_super) {
    __extends(UnparsableDeviceResponseError, _super);
    function UnparsableDeviceResponseError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, UnparsableDeviceResponseError.prototype);
        return _this;
    }
    return UnparsableDeviceResponseError;
}(Error));
exports.UnparsableDeviceResponseError = UnparsableDeviceResponseError;
var FailedDeviceResponseError = /** @class */ (function (_super) {
    __extends(FailedDeviceResponseError, _super);
    function FailedDeviceResponseError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, FailedDeviceResponseError.prototype);
        return _this;
    }
    return FailedDeviceResponseError;
}(Error));
exports.FailedDeviceResponseError = FailedDeviceResponseError;
var UnknownDeviceResponseError = /** @class */ (function (_super) {
    __extends(UnknownDeviceResponseError, _super);
    function UnknownDeviceResponseError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, UnknownDeviceResponseError.prototype);
        return _this;
    }
    return UnknownDeviceResponseError;
}(Error));
exports.UnknownDeviceResponseError = UnknownDeviceResponseError;
var CompileError = /** @class */ (function (_super) {
    __extends(CompileError, _super);
    function CompileError(message, results) {
        var _this = _super.call(this, message) || this;
        _this.results = results;
        Object.setPrototypeOf(_this, CompileError.prototype);
        return _this;
    }
    return CompileError;
}(Error));
exports.CompileError = CompileError;
var ConvertError = /** @class */ (function (_super) {
    __extends(ConvertError, _super);
    function ConvertError(message) {
        var _this = _super.call(this, message) || this;
        Object.setPrototypeOf(_this, ConvertError.prototype);
        return _this;
    }
    return ConvertError;
}(Error));
exports.ConvertError = ConvertError;
var MissingRequiredOptionError = /** @class */ (function (_super) {
    __extends(MissingRequiredOptionError, _super);
    function MissingRequiredOptionError(message) {
        var _this = _super.call(this, message) || this;
        Object.setPrototypeOf(_this, MissingRequiredOptionError.prototype);
        return _this;
    }
    return MissingRequiredOptionError;
}(Error));
exports.MissingRequiredOptionError = MissingRequiredOptionError;
