// brighterscript/dist/parser/tests/statement/Increment.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const Parser_1 = require("../../Parser");
const lexer_1 = require("../../../lexer");
const Parser_spec_1 = require("../Parser.spec");
const vscode_languageserver_1 = require("vscode-languageserver");
const DiagnosticMessages_1 = require("../../../DiagnosticMessages");
describe('parser postfix unary expressions', () => {
    it('parses postfix \'++\' for variables', () => {
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.identifier('foo'),
            Parser_spec_1.token(lexer_1.TokenKind.PlusPlus, '++'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.exist;
        chai_1.expect(statements).not.to.be.null;
    });
    it('parses postfix \'--\' for dotted get expressions', () => {
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.identifier('obj'),
            Parser_spec_1.token(lexer_1.TokenKind.Dot, '.'),
            Parser_spec_1.identifier('property'),
            Parser_spec_1.token(lexer_1.TokenKind.MinusMinus, '--'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.exist;
        chai_1.expect(statements).not.to.be.null;
    });
    it('parses postfix \'++\' for indexed get expressions', () => {
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.identifier('obj'),
            Parser_spec_1.token(lexer_1.TokenKind.LeftSquareBracket, '['),
            Parser_spec_1.identifier('property'),
            Parser_spec_1.token(lexer_1.TokenKind.RightSquareBracket, ']'),
            Parser_spec_1.token(lexer_1.TokenKind.PlusPlus, '++'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.exist;
        chai_1.expect(statements).not.to.be.null;
    });
    it('disallows consecutive postfix operators', () => {
        let { diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.identifier('foo'),
            Parser_spec_1.token(lexer_1.TokenKind.PlusPlus, '++'),
            Parser_spec_1.token(lexer_1.TokenKind.PlusPlus, '++'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(1);
        chai_1.expect(diagnostics[0]).deep.include({
            message: 'Consecutive increment/decrement operators are not allowed'
        });
    });
    it('disallows postfix \'--\' for function call results', () => {
        let { diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.identifier('func'),
            Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
            Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
            Parser_spec_1.token(lexer_1.TokenKind.MinusMinus, '--'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(1);
        chai_1.expect(diagnostics[0]).to.deep.include(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.incrementDecrementOperatorsAreNotAllowedAsResultOfFunctionCall()));
    });
    it('allows \'++\' at the end of a function', () => {
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.token(lexer_1.TokenKind.Sub, 'sub'),
            Parser_spec_1.identifier('foo'),
            Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
            Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            Parser_spec_1.identifier('someValue'),
            Parser_spec_1.token(lexer_1.TokenKind.PlusPlus, '++'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            Parser_spec_1.token(lexer_1.TokenKind.EndSub, 'end sub'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.exist;
        chai_1.expect(statements).not.to.be.null;
    });
    it('location tracking', () => {
        /**
         *    0   0   0   1
         *    0   4   8   2
         *  +--------------
         * 0| someNumber++
         */
        let { statements, diagnostics } = Parser_1.Parser.parse([
            {
                kind: lexer_1.TokenKind.Identifier,
                text: 'someNumber',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 0, 0, 10)
            },
            {
                kind: lexer_1.TokenKind.PlusPlus,
                text: '++',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 10, 0, 12)
            },
            {
                kind: lexer_1.TokenKind.Eof,
                text: '\0',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 12, 0, 13)
            }
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.be.lengthOf(1);
        chai_1.expect(statements[0].range).deep.include(vscode_languageserver_1.Range.create(0, 0, 0, 12));
    });
});
