// brighterscript/dist/parser/tests/controlFlow/For.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const Parser_1 = require("../../Parser");
const lexer_1 = require("../../../lexer");
const Parser_spec_1 = require("../Parser.spec");
const vscode_languageserver_1 = require("vscode-languageserver");
const __1 = require("../..");
describe('parser for loops', () => {
    it('accepts a \'step\' clause', () => {
        var _a;
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.token(lexer_1.TokenKind.For, 'for'),
            Parser_spec_1.identifier('i'),
            Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '0'),
            Parser_spec_1.token(lexer_1.TokenKind.To, 'to'),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
            Parser_spec_1.token(lexer_1.TokenKind.Step, 'step'),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '2'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            // body would go here, but it's not necessary for this test
            Parser_spec_1.token(lexer_1.TokenKind.EndFor, 'end for'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            Parser_spec_1.EOF
        ]);
        const statement = statements[0];
        chai_1.expect((_a = diagnostics[0]) === null || _a === void 0 ? void 0 : _a.message).not.to.exist;
        chai_1.expect(statement.increment).to.be.instanceof(__1.LiteralExpression);
        chai_1.expect(statement.increment.token.text).to.equal('2');
    });
    it('supports omitted \'step\' clause', () => {
        var _a;
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.token(lexer_1.TokenKind.For, 'for'),
            Parser_spec_1.identifier('i'),
            Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '0'),
            Parser_spec_1.token(lexer_1.TokenKind.To, 'to'),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            // body would go here, but it's not necessary for this test
            Parser_spec_1.token(lexer_1.TokenKind.EndFor, 'end for'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect((_a = diagnostics[0]) === null || _a === void 0 ? void 0 : _a.message).not.to.exist;
        chai_1.expect(statements[0].increment).not.to.exist;
    });
    it('allows \'next\' to terminate loop', () => {
        let { statements, diagnostics } = Parser_1.Parser.parse([
            Parser_spec_1.token(lexer_1.TokenKind.For, 'for'),
            Parser_spec_1.identifier('i'),
            Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '0'),
            Parser_spec_1.token(lexer_1.TokenKind.To, 'to'),
            Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            // body would go here, but it's not necessary for this test
            Parser_spec_1.token(lexer_1.TokenKind.Next, 'next'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.be.length.greaterThan(0);
    });
    it('supports a single trailing colon after the `to` expression', () => {
        var _a;
        const parser = Parser_1.Parser.parse(`
            for i = 1 to 3:
                print "for"
            end for
        `);
        chai_1.expect((_a = parser.diagnostics[0]) === null || _a === void 0 ? void 0 : _a.message).to.be.undefined;
    });
    it('location tracking', () => {
        /**
         *    0   0   0   1   1
         *    0   4   8   2   6
         *  +------------------
         * 0| for i = 0 to 10
         * 1|   Rnd(i)
         * 2| end for
         */
        let { statements, diagnostics } = Parser_1.Parser.parse([
            {
                kind: lexer_1.TokenKind.For,
                text: 'for',
                isReserved: true,
                range: vscode_languageserver_1.Range.create(0, 0, 0, 3)
            },
            {
                kind: lexer_1.TokenKind.Identifier,
                text: 'i',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 4, 0, 5)
            },
            {
                kind: lexer_1.TokenKind.Equal,
                text: '=',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 6, 0, 7)
            },
            {
                kind: lexer_1.TokenKind.IntegerLiteral,
                text: '0',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 8, 0, 9)
            },
            {
                kind: lexer_1.TokenKind.To,
                text: 'to',
                isReserved: false,
                range: {
                    start: { line: 0, column: 10 },
                    end: { start: 0, column: 12 }
                }
            },
            {
                kind: lexer_1.TokenKind.IntegerLiteral,
                text: '10',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 13, 0, 15)
            },
            {
                kind: lexer_1.TokenKind.Newline,
                text: '\n',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 15, 0, 16)
            },
            // loop body isn't significant for location tracking, so helper functions are safe
            Parser_spec_1.identifier('Rnd'),
            Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
            Parser_spec_1.identifier('i'),
            Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
            Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
            {
                kind: lexer_1.TokenKind.EndFor,
                text: 'end for',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(2, 0, 2, 8)
            },
            Parser_spec_1.EOF
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.be.lengthOf(1);
        chai_1.expect(statements[0].range).to.deep.include(vscode_languageserver_1.Range.create(0, 0, 2, 8));
    });
});
