// brighterscript/dist/parser/tests/expression/Function.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const Parser_1 = require("../../Parser");
const lexer_1 = require("../../../lexer");
const Parser_spec_1 = require("../Parser.spec");
const vscode_languageserver_1 = require("vscode-languageserver");
describe('parser', () => {
    describe('function expressions', () => {
        it('parses minimal empty function expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses colon-separated function declarations', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Colon, ':'),
                Parser_spec_1.token(lexer_1.TokenKind.Print, 'print'),
                Parser_spec_1.token(lexer_1.TokenKind.StringLiteral, 'Lorem ipsum'),
                Parser_spec_1.token(lexer_1.TokenKind.Colon, ':'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses non-empty function expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.Print, 'print'),
                Parser_spec_1.token(lexer_1.TokenKind.StringLiteral, 'Lorem ipsum'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses functions with implicit-dynamic arguments', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses functions with typed arguments', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('str'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('string'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('count'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('separator'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('object'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses functions with default argument expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '3'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '4'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('c'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Plus, '+'),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses functions with typed arguments and default expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '3'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Plus, '+'),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses return types', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('void'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
    });
    describe('sub expressions', () => {
        it('parses minimal sub expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Sub, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndSub, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses non-empty sub expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Sub, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.Print, 'print'),
                Parser_spec_1.token(lexer_1.TokenKind.StringLiteral, 'Lorem ipsum'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndSub, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses subs with implicit-dynamic arguments', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses subs with typed arguments', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('str'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('string'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('count'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('cb'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses subs with default argument expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Sub, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '3'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '4'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('c'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Plus, '+'),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndSub, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('parses subs with typed arguments and default expressions', () => {
            let { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('_'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Sub, 'sub'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '3'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.Comma, ','),
                Parser_spec_1.identifier('b'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.identifier('a'),
                Parser_spec_1.token(lexer_1.TokenKind.Plus, '+'),
                Parser_spec_1.token(lexer_1.TokenKind.IntegerLiteral, '5'),
                Parser_spec_1.token(lexer_1.TokenKind.As, 'as'),
                Parser_spec_1.identifier('integer'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndSub, 'end sub'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
    });
    describe('usage', () => {
        it('allows sub expressions in call arguments', () => {
            const { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('acceptsCallback'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.Print, 'print'),
                Parser_spec_1.token(lexer_1.TokenKind.StringLiteral, 'I\'m a callback'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
        it('allows function expressions in assignment RHS', () => {
            const { statements, diagnostics } = Parser_1.Parser.parse([
                Parser_spec_1.identifier('anonymousFunction'),
                Parser_spec_1.token(lexer_1.TokenKind.Equal, '='),
                Parser_spec_1.token(lexer_1.TokenKind.Function, 'function'),
                Parser_spec_1.token(lexer_1.TokenKind.LeftParen, '('),
                Parser_spec_1.token(lexer_1.TokenKind.RightParen, ')'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.Print, 'print'),
                Parser_spec_1.token(lexer_1.TokenKind.StringLiteral, 'I\'m anonymous'),
                Parser_spec_1.token(lexer_1.TokenKind.Newline, '\\n'),
                Parser_spec_1.token(lexer_1.TokenKind.EndFunction, 'end function'),
                Parser_spec_1.EOF
            ]);
            chai_1.expect(diagnostics).to.be.lengthOf(0);
            chai_1.expect(statements).to.be.length.greaterThan(0);
        });
    });
    it('location tracking', () => {
        /**
         *    0   0   0   1   1
         *    0   4   8   2   6
         *  +------------------
         * 0| _ = sub foo()
         * 1|
         * 2| end sub
         */
        let { statements, diagnostics } = Parser_1.Parser.parse([
            {
                kind: lexer_1.TokenKind.Identifier,
                text: '_',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 0, 0, 1)
            },
            {
                kind: lexer_1.TokenKind.Equal,
                text: '=',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 2, 0, 3)
            },
            {
                kind: lexer_1.TokenKind.Sub,
                text: 'sub',
                isReserved: true,
                range: vscode_languageserver_1.Range.create(0, 4, 0, 7)
            },
            {
                kind: lexer_1.TokenKind.LeftParen,
                text: '(',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 11, 0, 12)
            },
            {
                kind: lexer_1.TokenKind.RightParen,
                text: ')',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 12, 0, 13)
            },
            {
                kind: lexer_1.TokenKind.Newline,
                text: '\n',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(0, 13, 0, 14)
            },
            {
                kind: lexer_1.TokenKind.EndSub,
                text: 'end sub',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(2, 0, 2, 7)
            },
            {
                kind: lexer_1.TokenKind.Eof,
                text: '\0',
                isReserved: false,
                range: vscode_languageserver_1.Range.create(2, 7, 2, 8)
            }
        ]);
        chai_1.expect(diagnostics).to.be.lengthOf(0);
        chai_1.expect(statements).to.be.lengthOf(1);
        chai_1.expect(statements[0].value.range).to.deep.include(vscode_languageserver_1.Range.create(0, 4, 2, 7));
    });
});
