// brighterscript/dist/FunctionScope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionScope = void 0;
//TODO I think this class can be eliminated in favor of moving some of these onto the FunctionExpression AST node
class FunctionScope {
    constructor(func) {
        this.func = func;
        /**
         * The scopes that are children of this scope
         */
        this.childrenScopes = [];
        this.variableDeclarations = [];
    }
    /**
     * The full range of this function. Starts at the position of the `f` in function or `s` in sub,
     * and ends after the final `n` in `end function` or `b` in end sub.
     */
    get range() {
        var _a;
        return (_a = this.func) === null || _a === void 0 ? void 0 : _a.range;
    }
    /**
     * Find all variable declarations above the given line index
     * @param lineIndex
     */
    getVariablesAbove(lineIndex) {
        let results = [];
        for (let variableDeclaration of this.variableDeclarations) {
            if (variableDeclaration.lineIndex < lineIndex) {
                results.push(variableDeclaration);
            }
            else {
                break;
            }
        }
        return results;
    }
    getVariableByName(name) {
        name = name.toLowerCase();
        for (let variableDeclaration of this.variableDeclarations) {
            if (variableDeclaration.name.toLowerCase() === name) {
                return variableDeclaration;
            }
        }
    }
}
exports.FunctionScope = FunctionScope;
