// brighterscript/dist/ProgramBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgramBuilder = void 0;
const debounce = require("debounce-promise");
const path = require("path");
const rokuDeploy = require("roku-deploy");
const Program_1 = require("./Program");
const util_1 = require("./util");
const Watcher_1 = require("./Watcher");
const vscode_languageserver_1 = require("vscode-languageserver");
const Logger_1 = require("./Logger");
const PluginInterface_1 = require("./PluginInterface");
const diagnosticUtils = require("./diagnosticUtils");
/**
 * A runner class that handles
 */
class ProgramBuilder {
    constructor() {
        /**
         * Determines whether the console should be cleared after a run (true for cli, false for languageserver)
         */
        this.allowConsoleClearing = true;
        this.isRunning = false;
        this.logger = new Logger_1.Logger();
        this.plugins = new PluginInterface_1.default([], this.logger);
        this.fileResolvers = [];
        /**
         * A list of diagnostics that are always added to the `getDiagnostics()` call.
         */
        this.staticDiagnostics = [];
        /**
         * A method that is used to cancel a previous run task.
         * Does nothing if previous run has completed or was already canceled
         */
        this.cancelLastRun = () => {
            return Promise.resolve();
        };
    }
    addFileResolver(fileResolver) {
        this.fileResolvers.push(fileResolver);
        if (this.program) {
            this.program.fileResolvers.push(fileResolver);
        }
    }
    addDiagnostic(filePathAbsolute, diagnostic) {
        let file = this.program.getFileByPathAbsolute(filePathAbsolute);
        if (!file) {
            file = {
                pkgPath: this.program.getPkgPath(filePathAbsolute),
                pathAbsolute: filePathAbsolute,
                getDiagnostics: () => {
                    return [diagnostic];
                }
            };
        }
        diagnostic.file = file;
        this.staticDiagnostics.push(diagnostic);
    }
    getDiagnostics() {
        return [
            ...this.staticDiagnostics,
            ...this.program ? this.program.getDiagnostics() : []
        ];
    }
    async run(options) {
        this.logger.logLevel = options.logLevel;
        if (this.isRunning) {
            throw new Error('Server is already running');
        }
        this.isRunning = true;
        try {
            this.options = await util_1.util.normalizeAndResolveConfig(options);
            this.loadPlugins();
        }
        catch (e) {
            if ((e === null || e === void 0 ? void 0 : e.file) && e.message && e.code) {
                let err = e;
                this.staticDiagnostics.push(err);
            }
            else {
                //if this is not a diagnostic, something else is wrong...
                throw e;
            }
            this.printDiagnostics();
            //we added diagnostics, so hopefully that draws attention to the underlying issues.
            //For now, just use a default options object so we have a functioning program
            this.options = util_1.util.normalizeConfig({});
        }
        this.logger.logLevel = this.options.logLevel;
        this.program = this.createProgram();
        //parse every file in the entire project
        this.logger.log('Parsing files');
        await this.loadAllFilesAST();
        if (this.options.watch) {
            this.logger.log('Starting compilation in watch mode...');
            await this.runOnce();
            this.enableWatchMode();
        }
        else {
            await this.runOnce();
        }
    }
    createProgram() {
        const program = new Program_1.Program(this.options, undefined, this.plugins);
        //add the initial FileResolvers
        program.fileResolvers.push(...this.fileResolvers);
        this.plugins.emit('afterProgramCreate', program);
        return program;
    }
    loadPlugins() {
        var _a;
        util_1.loadPlugins((_a = this.options.cwd) !== null && _a !== void 0 ? _a : process.cwd(), this.options.plugins, (pathOrModule, err) => this.logger.error(`Error when loading plugin '${pathOrModule}':`, err)).forEach(plugin => this.plugins.add(plugin));
        this.plugins.emit('beforeProgramCreate', this);
    }
    clearConsole() {
        if (this.allowConsoleClearing) {
            util_1.util.clearConsole();
        }
    }
    enableWatchMode() {
        this.watcher = new Watcher_1.Watcher(this.options);
        //keep the process alive indefinitely by setting an interval that runs once every 12 days
        setInterval(() => { }, 1073741824);
        //clear the console
        this.clearConsole();
        let fileObjects = rokuDeploy.normalizeFilesArray(this.options.files ? this.options.files : []);
        //add each set of files to the file watcher
        for (let fileObject of fileObjects) {
            let src = typeof fileObject === 'string' ? fileObject : fileObject.src;
            this.watcher.watch(src);
        }
        this.logger.log('Watching for file changes...');
        let debouncedRunOnce = debounce(async () => {
            this.logger.log('File change detected. Starting incremental compilation...');
            await this.runOnce();
            let errorCount = this.getDiagnostics().length;
            this.logger.log(`Found ${errorCount} errors. Watching for file changes.`);
        }, 50);
        //on any file watcher event
        this.watcher.on('all', async (event, thePath) => {
            thePath = util_1.standardizePath `${path.resolve(this.rootDir, thePath)}`;
            if (event === 'add' || event === 'change') {
                await this.program.addOrReplaceFile({
                    src: thePath,
                    dest: rokuDeploy.getDestPath(thePath, this.program.options.files, this.rootDir)
                });
            }
            else if (event === 'unlink') {
                this.program.removeFile(thePath);
            }
            //wait for change events to settle, and then execute `run`
            await debouncedRunOnce();
        });
    }
    /**
     * The rootDir for this program.
     */
    get rootDir() {
        return this.program.options.rootDir;
    }
    /**
     * Run the entire process exactly one time.
     */
    runOnce() {
        //clear the console
        this.clearConsole();
        let cancellationToken = { isCanceled: false };
        //wait for the previous run to complete
        let runPromise = this.cancelLastRun().then(() => {
            //start the new run
            return this._runOnce(cancellationToken);
        });
        //a function used to cancel this run
        this.cancelLastRun = () => {
            cancellationToken.isCanceled = true;
            return runPromise;
        };
        return runPromise;
    }
    printDiagnostics() {
        var _a, _b;
        if (this.options.showDiagnosticsInConsole === false) {
            return;
        }
        let diagnostics = this.getDiagnostics();
        //group the diagnostics by file
        let diagnosticsByFile = {};
        for (let diagnostic of diagnostics) {
            if (!diagnosticsByFile[diagnostic.file.pathAbsolute]) {
                diagnosticsByFile[diagnostic.file.pathAbsolute] = [];
            }
            diagnosticsByFile[diagnostic.file.pathAbsolute].push(diagnostic);
        }
        //get printing options
        const options = diagnosticUtils.getPrintDiagnosticOptions(this.options);
        const { cwd, emitFullPaths } = options;
        let pathsAbsolute = Object.keys(diagnosticsByFile).sort();
        for (let pathAbsolute of pathsAbsolute) {
            let diagnosticsForFile = diagnosticsByFile[pathAbsolute];
            //sort the diagnostics in line and column order
            let sortedDiagnostics = diagnosticsForFile.sort((a, b) => {
                return (a.range.start.line - b.range.start.line ||
                    a.range.start.character - b.range.start.character);
            });
            let filePath = pathAbsolute;
            if (!emitFullPaths) {
                filePath = path.relative(cwd, filePath);
            }
            //load the file text
            const file = this.program.getFileByPathAbsolute(pathAbsolute);
            //get the file's in-memory contents if available
            const lines = (_b = (_a = file === null || file === void 0 ? void 0 : file.fileContents) === null || _a === void 0 ? void 0 : _a.split(/\r?\n/g)) !== null && _b !== void 0 ? _b : [];
            for (let diagnostic of sortedDiagnostics) {
                //default the severity to error if undefined
                let severity = typeof diagnostic.severity === 'number' ? diagnostic.severity : vscode_languageserver_1.DiagnosticSeverity.Error;
                //format output
                diagnosticUtils.printDiagnostic(options, severity, filePath, lines, diagnostic);
            }
        }
    }
    /**
     * Run the process once, allowing cancelability.
     * NOTE: This should only be called by `runOnce`.
     * @param cancellationToken
     */
    async _runOnce(cancellationToken) {
        let wereDiagnosticsPrinted = false;
        try {
            //maybe cancel?
            if (cancellationToken.isCanceled === true) {
                return -1;
            }
            this.logger.log('Validating project');
            //validate program
            await this.validateProject();
            //maybe cancel?
            if (cancellationToken.isCanceled === true) {
                return -1;
            }
            this.printDiagnostics();
            wereDiagnosticsPrinted = true;
            let errorCount = this.getDiagnostics().filter(x => x.severity === vscode_languageserver_1.DiagnosticSeverity.Error).length;
            if (errorCount > 0) {
                this.logger.log(`Found ${errorCount} ${errorCount === 1 ? 'error' : 'errors'}`);
                return errorCount;
            }
            //create the deployment package (and transpile as well)
            await this.createPackageIfEnabled();
            //maybe cancel?
            if (cancellationToken.isCanceled === true) {
                return -1;
            }
            //deploy the package
            await this.deployPackageIfEnabled();
            return 0;
        }
        catch (e) {
            if (wereDiagnosticsPrinted === false) {
                this.printDiagnostics();
            }
            throw e;
        }
    }
    async createPackageIfEnabled() {
        if (this.options.copyToStaging || this.options.createPackage || this.options.deploy) {
            //transpile the project
            await this.transpile();
            //create the zip file if configured to do so
            if (this.options.createPackage !== false || this.options.deploy) {
                await this.logger.time(Logger_1.LogLevel.log, [`Creating package at ${this.options.outFile}`], async () => {
                    await rokuDeploy.zipPackage(Object.assign(Object.assign({}, this.options), { outDir: util_1.util.getOutDir(this.options), outFile: path.basename(this.options.outFile) }));
                });
            }
        }
    }
    /**
     * Transpiles the entire program into the staging folder
     */
    async transpile() {
        let options = util_1.util.cwdWork(this.options.cwd, () => {
            return rokuDeploy.getOptions(Object.assign(Object.assign({}, this.options), { outDir: util_1.util.getOutDir(this.options), outFile: path.basename(this.options.outFile) }));
        });
        //get every file referenced by the files array
        let fileMap = await rokuDeploy.getFilePaths(options.files, options.rootDir);
        //remove files currently loaded in the program, we will transpile those instead (even if just for source maps)
        let filteredFileMap = [];
        for (let fileEntry of fileMap) {
            if (this.program.hasFile(fileEntry.src) === false) {
                filteredFileMap.push(fileEntry);
            }
        }
        this.plugins.emit('beforePrepublish', this, filteredFileMap);
        await this.logger.time(Logger_1.LogLevel.log, ['Copying to staging directory'], async () => {
            //prepublish all non-program-loaded files to staging
            await rokuDeploy.prepublishToStaging(Object.assign(Object.assign({}, options), { files: filteredFileMap }));
        });
        this.plugins.emit('afterPrepublish', this, filteredFileMap);
        this.plugins.emit('beforePublish', this, fileMap);
        await this.logger.time(Logger_1.LogLevel.log, ['Transpiling'], async () => {
            //transpile any brighterscript files
            await this.program.transpile(fileMap, options.stagingFolderPath);
        });
        this.plugins.emit('afterPublish', this, fileMap);
    }
    async deployPackageIfEnabled() {
        //deploy the project if configured to do so
        if (this.options.deploy) {
            await this.logger.time(Logger_1.LogLevel.log, ['Deploying package to', this.options.host], async () => {
                await rokuDeploy.publish(Object.assign(Object.assign({}, this.options), { outDir: util_1.util.getOutDir(this.options), outFile: path.basename(this.options.outFile) }));
            });
        }
    }
    /**
     * Parse and load the AST for every file in the project
     */
    async loadAllFilesAST() {
        let errorCount = 0;
        let files = await this.logger.time(Logger_1.LogLevel.debug, ['ProgramBuilder.loadAllFilesAST()'], async () => {
            return util_1.util.getFilePaths(this.options);
        });
        this.logger.debug('ProgramBuilder.loadAllFilesAST() files:', files);
        const typedefFiles = [];
        const nonTypedefFiles = [];
        for (const file of files) {
            const srcLower = file.src.toLowerCase();
            if (srcLower.endsWith('.d.bs')) {
                typedefFiles.push(file);
            }
            else {
                nonTypedefFiles.push(file);
            }
        }
        //preload every type definition file first, which eliminates duplicate file loading
        await Promise.all(typedefFiles.map(async (file) => {
            try {
                await this.program.addOrReplaceFile(file);
            }
            catch (e) {
                //log the error, but don't fail this process because the file might be fixable later
                this.logger.log(e);
            }
        }));
        //parse every file other than the type definitions
        await Promise.all(nonTypedefFiles.map(async (file) => {
            try {
                let fileExtension = path.extname(file.src).toLowerCase();
                //only process certain file types
                if (['.bs', '.brs', '.xml'].includes(fileExtension)) {
                    await this.program.addOrReplaceFile(file);
                }
            }
            catch (e) {
                //log the error, but don't fail this process because the file might be fixable later
                this.logger.log(e);
            }
        }));
        return errorCount;
    }
    /**
     * Remove all files from the program that are in the specified folder path
     * @param folderPathAbsolute
     */
    removeFilesInFolder(folderPathAbsolute) {
        for (let filePath in this.program.files) {
            //if the file path starts with the parent path and the file path does not exactly match the folder path
            if (filePath.startsWith(folderPathAbsolute) && filePath !== folderPathAbsolute) {
                this.program.removeFile(filePath);
            }
        }
    }
    /**
     * Scan every file and resolve all variable references.
     * If no errors were encountered, return true. Otherwise return false.
     */
    async validateProject() {
        await this.program.validate();
    }
    dispose() {
        var _a, _b;
        if (this.watcher) {
            this.watcher.dispose();
        }
        if (this.program) {
            (_b = (_a = this.program).dispose) === null || _b === void 0 ? void 0 : _b.call(_a);
        }
    }
}
exports.ProgramBuilder = ProgramBuilder;
