// brighterscript/dist/astUtils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiagnosticTag = exports.DiagnosticSeverity = exports.CancellationTokenSource = exports.CancellationToken = exports.Position = exports.Range = void 0;
const vscode_languageserver_1 = require("vscode-languageserver");
Object.defineProperty(exports, "Range", { enumerable: true, get: function () { return vscode_languageserver_1.Range; } });
Object.defineProperty(exports, "Position", { enumerable: true, get: function () { return vscode_languageserver_1.Position; } });
Object.defineProperty(exports, "CancellationToken", { enumerable: true, get: function () { return vscode_languageserver_1.CancellationToken; } });
Object.defineProperty(exports, "CancellationTokenSource", { enumerable: true, get: function () { return vscode_languageserver_1.CancellationTokenSource; } });
Object.defineProperty(exports, "DiagnosticSeverity", { enumerable: true, get: function () { return vscode_languageserver_1.DiagnosticSeverity; } });
Object.defineProperty(exports, "DiagnosticTag", { enumerable: true, get: function () { return vscode_languageserver_1.DiagnosticTag; } });
__exportStar(require("./visitors"), exports);
__exportStar(require("./stackedVisitor"), exports);
__exportStar(require("./reflection"), exports);
__exportStar(require("./creators"), exports);
