// brighterscript/dist/validators/ClassValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BsClassValidator = void 0;
const DiagnosticMessages_1 = require("../DiagnosticMessages");
const Parser_1 = require("../parser/Parser");
const vscode_languageserver_1 = require("vscode-languageserver");
const vscode_uri_1 = require("vscode-uri");
const util_1 = require("../util");
const reflection_1 = require("../astUtils/reflection");
const astUtils_1 = require("../astUtils");
class BsClassValidator {
    validate(scope) {
        this.scope = scope;
        this.diagnostics = [];
        this.classes = {};
        this.findClasses();
        this.findNamespaceNonNamespaceCollisions();
        this.linkClassesWithParents();
        this.validateMemberCollisions();
        this.verifyChildConstructor();
        this.verifyNewExpressions();
        this.cleanUp();
    }
    /**
     * Given a class name optionally prefixed with a namespace name, find the class that matches
     */
    getClassByName(className, namespaceName) {
        let fullName = util_1.default.getFullyQualifiedClassName(className, namespaceName);
        let cls = this.classes[fullName.toLowerCase()];
        //if we couldn't find the class by its full namespaced name, look for a global class with that name
        if (!cls) {
            cls = this.classes[className.toLowerCase()];
        }
        return cls;
    }
    /**
     * Find all "new" statements in the program,
     * and make sure we can find a class with that name
     */
    verifyNewExpressions() {
        var _a, _b;
        let files = this.scope.getFiles();
        for (let file of files) {
            let newExpressions = file.parser.references.newExpressions;
            for (let newExpression of newExpressions) {
                let className = newExpression.className.getName(Parser_1.ParseMode.BrighterScript);
                let newableClass = this.getClassByName(className, (_a = newExpression.namespaceName) === null || _a === void 0 ? void 0 : _a.getName(Parser_1.ParseMode.BrighterScript));
                if (!newableClass) {
                    //try and find functions with this name.
                    let fullName = util_1.default.getFullyQualifiedClassName(className, (_b = newExpression.namespaceName) === null || _b === void 0 ? void 0 : _b.getName(Parser_1.ParseMode.BrighterScript));
                    let callable = this.scope.getCallableByName(fullName);
                    //if we found a callable with this name, the user used a "new" keyword in front of a function. add error
                    if (callable) {
                        this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.expressionIsNotConstructable(callable.isSub ? 'sub' : 'function')), { file: file, range: newExpression.className.range }));
                        //could not find a class with this name
                    }
                    else {
                        this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.classCouldNotBeFound(className, this.scope.name)), { file: file, range: newExpression.className.range }));
                    }
                }
            }
        }
    }
    findNamespaceNonNamespaceCollisions() {
        for (let name in this.classes) {
            let classStatement = this.classes[name];
            //catch namespace class collision with global class
            let nonNamespaceClass = this.classes[util_1.default.getTextAfterFinalDot(name).toLowerCase()];
            if (classStatement.namespaceName && nonNamespaceClass) {
                this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.namespacedClassCannotShareNamewithNonNamespacedClass(nonNamespaceClass.name.text)), { file: classStatement.file, range: classStatement.name.range, relatedInformation: [{
                            location: vscode_languageserver_1.Location.create(vscode_uri_1.URI.file(nonNamespaceClass.file.pathAbsolute).toString(), nonNamespaceClass.name.range),
                            message: 'Original class declared here'
                        }] }));
            }
        }
    }
    verifyChildConstructor() {
        for (let key in this.classes) {
            let classStatement = this.classes[key];
            const newMethod = classStatement.memberMap.new;
            if (
            //this class has a "new method"
            newMethod &&
                //this class has a parent class
                classStatement.parentClass) {
                //prevent use of `m.` anywhere before the `super()` call
                const cancellationToken = new vscode_languageserver_1.CancellationTokenSource();
                let superCall;
                newMethod.func.body.walk(astUtils_1.createVisitor({
                    VariableExpression: (expression, parent) => {
                        var _a;
                        const expressionNameLower = (_a = expression === null || expression === void 0 ? void 0 : expression.name) === null || _a === void 0 ? void 0 : _a.text.toLowerCase();
                        if (expressionNameLower === 'm') {
                            this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.classConstructorIllegalUseOfMBeforeSuperCall()), { file: classStatement.file, range: expression.range }));
                        }
                        if (reflection_1.isCallExpression(parent) && expressionNameLower === 'super') {
                            superCall = parent;
                            //stop walking
                            cancellationToken.cancel();
                        }
                    }
                }), {
                    walkMode: astUtils_1.WalkMode.visitAll,
                    cancel: cancellationToken.token
                });
                //every child class constructor must include a call to `super()` (except for typedef files)
                if (!superCall && !classStatement.file.isTypedef) {
                    this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.classConstructorMissingSuperCall()), { file: classStatement.file, range: newMethod.range }));
                }
            }
        }
    }
    validateMemberCollisions() {
        for (let key in this.classes) {
            let classStatement = this.classes[key];
            let methods = {};
            let fields = {};
            for (let statement of classStatement.body) {
                if (reflection_1.isClassMethodStatement(statement) || reflection_1.isClassFieldStatement(statement)) {
                    let member = statement;
                    let lowerMemberName = member.name.text.toLowerCase();
                    //catch duplicate member names on same class
                    if (methods[lowerMemberName] || fields[lowerMemberName]) {
                        this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.duplicateIdentifier(member.name.text)), { file: classStatement.file, range: member.name.range }));
                    }
                    let memberType = reflection_1.isClassFieldStatement(member) ? 'field' : 'method';
                    let ancestorAndMember = this.getAncestorMember(classStatement, lowerMemberName);
                    if (ancestorAndMember) {
                        let ancestorMemberType = reflection_1.isClassFieldStatement(ancestorAndMember.member) ? 'field' : 'method';
                        //mismatched member type (field/method in child, opposite in parent)
                        if (memberType !== ancestorMemberType) {
                            this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.classChildMemberDifferentMemberTypeThanAncestor(memberType, ancestorMemberType, ancestorAndMember.classStatement.getName(Parser_1.ParseMode.BrighterScript))), { file: classStatement.file, range: member.range }));
                        }
                        //child field has same name as parent
                        if (reflection_1.isClassFieldStatement(member)) {
                            this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.memberAlreadyExistsInParentClass(memberType, ancestorAndMember.classStatement.getName(Parser_1.ParseMode.BrighterScript))), { file: classStatement.file, range: member.range }));
                        }
                        //child method missing the override keyword
                        if (
                        //is a method
                        reflection_1.isClassMethodStatement(member) &&
                            //does not have an override keyword
                            !member.override &&
                            //is not the constructur function
                            member.name.text.toLowerCase() !== 'new') {
                            this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.missingOverrideKeyword(ancestorAndMember.classStatement.getName(Parser_1.ParseMode.BrighterScript))), { file: classStatement.file, range: member.range }));
                        }
                    }
                    if (reflection_1.isClassMethodStatement(member)) {
                        methods[lowerMemberName] = member;
                    }
                    else if (reflection_1.isClassFieldStatement(member)) {
                        fields[lowerMemberName] = member;
                    }
                }
            }
        }
    }
    /**
     * Get the closest member with the specified name (case-insensitive)
     */
    getAncestorMember(classStatement, memberName) {
        let lowerMemberName = memberName.toLowerCase();
        let ancestor = classStatement.parentClass;
        while (ancestor) {
            let member = ancestor.memberMap[lowerMemberName];
            if (member) {
                return {
                    member: member,
                    classStatement: ancestor
                };
            }
            ancestor = ancestor.parentClass;
        }
    }
    cleanUp() {
        //unlink all classes from their parents so it doesn't mess up the next scope
        for (let key in this.classes) {
            let classStatement = this.classes[key];
            delete classStatement.parentClass;
            delete classStatement.file;
        }
    }
    findClasses() {
        var _a;
        this.classes = {};
        let files = this.scope.getFiles();
        for (let file of files) {
            for (let x of (_a = file.parser.references.classStatements) !== null && _a !== void 0 ? _a : []) {
                let classStatement = x;
                let name = classStatement.getName(Parser_1.ParseMode.BrighterScript);
                //skip this class if it doesn't have a name
                if (!name) {
                    continue;
                }
                let lowerName = name.toLowerCase();
                //see if this class was already defined
                let alreadyDefinedClass = this.classes[lowerName];
                //if we don't already have this class, register it
                if (!alreadyDefinedClass) {
                    this.classes[lowerName] = classStatement;
                    classStatement.file = file;
                    //add a diagnostic about this class already existing
                }
                else {
                    this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.duplicateClassDeclaration(this.scope.name, name)), { file: file, range: classStatement.name.range, relatedInformation: [{
                                location: vscode_languageserver_1.Location.create(vscode_uri_1.URI.file(alreadyDefinedClass.file.pathAbsolute).toString(), this.classes[lowerName].range),
                                message: ''
                            }] }));
                }
            }
        }
    }
    linkClassesWithParents() {
        var _a;
        //link all classes with their parents
        for (let key in this.classes) {
            let classStatement = this.classes[key];
            let parentClassName = (_a = classStatement.parentClassName) === null || _a === void 0 ? void 0 : _a.getName(Parser_1.ParseMode.BrighterScript);
            if (parentClassName) {
                let relativeName;
                let absoluteName;
                //if the parent class name was namespaced in the declaration of this class,
                //compute the relative name of the parent class and the absolute name of the parent class
                if (parentClassName.indexOf('.') > 0) {
                    absoluteName = parentClassName;
                    let parts = parentClassName.split('.');
                    relativeName = parts[parts.length - 1];
                    //the parent class name was NOT namespaced.
                    //compute the relative name of the parent class and prepend the current class's namespace
                    //to the beginning of the parent class's name
                }
                else {
                    if (classStatement.namespaceName) {
                        absoluteName = `${classStatement.namespaceName.getName(Parser_1.ParseMode.BrighterScript)}.${parentClassName}`;
                    }
                    else {
                        absoluteName = parentClassName;
                    }
                    relativeName = parentClassName;
                }
                let relativeParent = this.classes[relativeName.toLowerCase()];
                let absoluteParent = this.classes[absoluteName.toLowerCase()];
                let parentClass;
                //if we found a relative parent class
                if (relativeParent) {
                    parentClass = relativeParent;
                    //we found an absolute parent class
                }
                else if (absoluteParent) {
                    parentClass = absoluteParent;
                    //couldn't find the parent class
                }
                else {
                    this.diagnostics.push(Object.assign(Object.assign({}, DiagnosticMessages_1.DiagnosticMessages.classCouldNotBeFound(parentClassName, this.scope.name)), { file: classStatement.file, range: classStatement.parentClassName.range }));
                }
                classStatement.parentClass = parentClass;
            }
        }
    }
}
exports.BsClassValidator = BsClassValidator;
