// brighterscript/dist/parser/tests/Parser.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EOF = exports.identifier = exports.token = void 0;
const lexer_1 = require("../../lexer");
const creators_1 = require("../../astUtils/creators");
/* A set of utilities to be used while writing tests for the BRS parser. */
/**
 * Creates a token with the given `kind` and (optional) `literal` value.
 * @param {TokenKind} kind the tokenKind the produced token should represent.
 * @param {string} text the text represented by this token.
 * @param {*} [literal] the literal value that the produced token should contain, if any
 * @returns {object} a token of `kind` representing `text` with value `literal`.
 */
function token(kind, text) {
    return {
        kind: kind,
        text: text,
        isReserved: lexer_1.ReservedWords.has((text || '').toLowerCase()),
        range: creators_1.interpolatedRange,
        leadingWhitespace: ''
    };
}
exports.token = token;
/**
 * Creates an Identifier token with the given `text`.
 * @param {string} text
 * @returns {object} a token with the provided `text`.
 */
function identifier(text) {
    return exports.token(lexer_1.TokenKind.Identifier, text);
}
exports.identifier = identifier;
/** An end-of-file token. */
exports.EOF = token(lexer_1.TokenKind.Eof, '\0');
