// brighterscript/dist/parser/Expression.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnnotationExpression = exports.TaggedTemplateStringExpression = exports.TemplateStringExpression = exports.TemplateStringQuasiExpression = exports.CallfuncExpression = exports.NewExpression = exports.SourceLiteralExpression = exports.VariableExpression = exports.UnaryExpression = exports.AALiteralExpression = exports.AAMemberExpression = exports.ArrayLiteralExpression = exports.EscapedCharCodeLiteralExpression = exports.LiteralExpression = exports.GroupingExpression = exports.IndexedGetExpression = exports.XmlAttributeGetExpression = exports.DottedGetExpression = exports.NamespacedVariableNameExpression = exports.FunctionParameterExpression = exports.FunctionExpression = exports.CallExpression = exports.BinaryExpression = exports.Expression = void 0;
const lexer_1 = require("../lexer");
const source_map_1 = require("source-map");
const util_1 = require("../util");
const Parser_1 = require("./Parser");
const fileUrl = require("file-url");
const visitors_1 = require("../astUtils/visitors");
const reflection_1 = require("../astUtils/reflection");
const VoidType_1 = require("../types/VoidType");
const DynamicType_1 = require("../types/DynamicType");
/** A BrightScript expression */
class Expression {
    constructor() {
        /**
         * When being considered by the walk visitor, this describes what type of element the current class is.
         */
        this.visitMode = visitors_1.InternalWalkMode.visitExpressions;
    }
}
exports.Expression = Expression;
class BinaryExpression extends Expression {
    constructor(left, operator, right) {
        super();
        this.left = left;
        this.operator = operator;
        this.right = right;
        this.range = util_1.default.createRangeFromPositions(this.left.range.start, this.right.range.end);
    }
    transpile(state) {
        return [
            new source_map_1.SourceNode(this.left.range.start.line + 1, this.left.range.start.character, state.pathAbsolute, this.left.transpile(state)),
            ' ',
            new source_map_1.SourceNode(this.operator.range.start.line + 1, this.operator.range.start.character, state.pathAbsolute, this.operator.text),
            ' ',
            new source_map_1.SourceNode(this.right.range.start.line + 1, this.right.range.start.character, state.pathAbsolute, this.right.transpile(state))
        ];
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'left', visitor, options);
            visitors_1.walk(this, 'right', visitor, options);
        }
    }
}
exports.BinaryExpression = BinaryExpression;
class CallExpression extends Expression {
    constructor(callee, openingParen, closingParen, args, 
    /**
     * The namespace that currently wraps this call expression. This is NOT the namespace of the callee...that will be represented in the callee expression itself.
     */
    namespaceName) {
        super();
        this.callee = callee;
        this.openingParen = openingParen;
        this.closingParen = closingParen;
        this.args = args;
        this.namespaceName = namespaceName;
        this.range = util_1.default.createRangeFromPositions(this.callee.range.start, this.closingParen.range.end);
    }
    transpile(state, nameOverride) {
        let result = [];
        //transpile the name
        if (nameOverride) {
            result.push(state.sourceNode(this.callee, nameOverride));
        }
        else {
            result.push(...this.callee.transpile(state));
        }
        result.push(new source_map_1.SourceNode(this.openingParen.range.start.line + 1, this.openingParen.range.start.character, state.pathAbsolute, '('));
        for (let i = 0; i < this.args.length; i++) {
            //add comma between args
            if (i > 0) {
                result.push(', ');
            }
            let arg = this.args[i];
            result.push(...arg.transpile(state));
        }
        result.push(new source_map_1.SourceNode(this.closingParen.range.start.line + 1, this.closingParen.range.start.character, state.pathAbsolute, ')'));
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'callee', visitor, options);
            for (let i = 0; i < this.args.length; i++) {
                visitors_1.walk(this.args, i, visitor, options, this);
            }
        }
    }
}
exports.CallExpression = CallExpression;
CallExpression.MaximumArguments = 32;
class FunctionExpression extends Expression {
    constructor(parameters, body, functionType, end, leftParen, rightParen, asToken, returnTypeToken, 
    /**
     * If this function is enclosed within another function, this will reference that parent function
     */
    parentFunction) {
        super();
        this.parameters = parameters;
        this.body = body;
        this.functionType = functionType;
        this.end = end;
        this.leftParen = leftParen;
        this.rightParen = rightParen;
        this.asToken = asToken;
        this.returnTypeToken = returnTypeToken;
        this.parentFunction = parentFunction;
        /**
         * The list of function calls that are declared within this function scope. This excludes CallExpressions
         * declared in child functions
         */
        this.callExpressions = [];
        /**
         * A list of all child functions declared directly within this function
         */
        this.childFunctionExpressions = [];
        if (this.returnTypeToken) {
            this.returnType = util_1.default.tokenToBscType(this.returnTypeToken);
        }
        else if (this.functionType.text.toLowerCase() === 'sub') {
            this.returnType = new VoidType_1.VoidType();
        }
        else {
            this.returnType = new DynamicType_1.DynamicType();
        }
    }
    /**
     * The range of the function, starting at the 'f' in function or 's' in sub (or the open paren if the keyword is missing),
     * and ending with the last n' in 'end function' or 'b' in 'end sub'
     */
    get range() {
        var _a, _b, _c, _d, _e;
        return util_1.default.createRangeFromPositions(((_a = this.functionType) !== null && _a !== void 0 ? _a : this.leftParen).range.start, ((_e = (_d = (_c = (_b = this.end) !== null && _b !== void 0 ? _b : this.body) !== null && _c !== void 0 ? _c : this.returnTypeToken) !== null && _d !== void 0 ? _d : this.asToken) !== null && _e !== void 0 ? _e : this.rightParen).range.end);
    }
    transpile(state, name, includeBody = true) {
        let results = [];
        //'function'|'sub'
        results.push(new source_map_1.SourceNode(this.functionType.range.start.line + 1, this.functionType.range.start.character, state.pathAbsolute, this.functionType.text.toLowerCase()));
        //functionName?
        if (name) {
            results.push(' ', new source_map_1.SourceNode(name.range.start.line + 1, name.range.start.character, state.pathAbsolute, name.text));
        }
        //leftParen
        results.push(new source_map_1.SourceNode(this.leftParen.range.start.line + 1, this.leftParen.range.start.character, state.pathAbsolute, '('));
        //parameters
        for (let i = 0; i < this.parameters.length; i++) {
            let param = this.parameters[i];
            //add commas
            if (i > 0) {
                results.push(', ');
            }
            //add parameter
            results.push(param.transpile(state));
        }
        //right paren
        results.push(new source_map_1.SourceNode(this.rightParen.range.start.line + 1, this.rightParen.range.start.character, state.pathAbsolute, ')'));
        //as [Type]
        if (this.asToken) {
            results.push(' ', 
            //as
            new source_map_1.SourceNode(this.asToken.range.start.line + 1, this.asToken.range.start.character, state.pathAbsolute, 'as'), ' ', 
            //return type
            new source_map_1.SourceNode(this.returnTypeToken.range.start.line + 1, this.returnTypeToken.range.start.character, state.pathAbsolute, this.returnTypeToken.text.toLowerCase()));
        }
        if (includeBody) {
            state.lineage.unshift(this);
            let body = this.body.transpile(state);
            state.lineage.shift();
            results.push(...body);
        }
        results.push('\n');
        //'end sub'|'end function'
        results.push(state.indent(), new source_map_1.SourceNode(this.end.range.start.line + 1, this.end.range.start.character, state.pathAbsolute, this.end.text));
        return results;
    }
    getTypedef(state, name) {
        return this.transpile(state, name, false);
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            for (let i = 0; i < this.parameters.length; i++) {
                visitors_1.walk(this.parameters, i, visitor, options, this);
            }
            //This is the core of full-program walking...it allows us to step into sub functions
            if (options.walkMode & visitors_1.InternalWalkMode.recurseChildFunctions) {
                visitors_1.walk(this, 'body', visitor, options);
            }
        }
    }
}
exports.FunctionExpression = FunctionExpression;
class FunctionParameterExpression extends Expression {
    constructor(name, typeToken, defaultValue, asToken) {
        super();
        this.name = name;
        this.typeToken = typeToken;
        this.defaultValue = defaultValue;
        this.asToken = asToken;
        if (typeToken) {
            this.type = util_1.default.tokenToBscType(typeToken);
        }
        else {
            this.type = new DynamicType_1.DynamicType();
        }
    }
    get range() {
        return {
            start: this.name.range.start,
            end: this.typeToken ? this.typeToken.range.end : this.name.range.end
        };
    }
    transpile(state) {
        let result = [
            //name
            new source_map_1.SourceNode(this.name.range.start.line + 1, this.name.range.start.character, state.pathAbsolute, this.name.text)
        ];
        //default value
        if (this.defaultValue) {
            result.push(' = ');
            result.push(this.defaultValue.transpile(state));
        }
        //type declaration
        if (this.asToken) {
            result.push(' ');
            result.push(new source_map_1.SourceNode(this.asToken.range.start.line + 1, this.asToken.range.start.character, state.pathAbsolute, 'as'));
            result.push(' ');
            result.push(new source_map_1.SourceNode(this.typeToken.range.start.line + 1, this.typeToken.range.start.character, state.pathAbsolute, this.typeToken.text));
        }
        return result;
    }
    walk(visitor, options) {
        // eslint-disable-next-line no-bitwise
        if (this.defaultValue && options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'defaultValue', visitor, options);
        }
    }
}
exports.FunctionParameterExpression = FunctionParameterExpression;
class NamespacedVariableNameExpression extends Expression {
    constructor(
    //if this is a `DottedGetExpression`, it must be comprised only of `VariableExpression`s
    expression) {
        super();
        this.expression = expression;
        this.range = expression.range;
    }
    transpile(state) {
        return [
            new source_map_1.SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, this.getName(Parser_1.ParseMode.BrightScript))
        ];
    }
    getNameParts() {
        let parts = [];
        if (reflection_1.isVariableExpression(this.expression)) {
            parts.push(this.expression.name.text);
        }
        else {
            let expr = this.expression;
            parts.push(expr.name.text);
            while (reflection_1.isVariableExpression(expr) === false) {
                expr = expr.obj;
                parts.unshift(expr.name.text);
            }
        }
        return parts;
    }
    getName(parseMode) {
        if (parseMode === Parser_1.ParseMode.BrighterScript) {
            return this.getNameParts().join('.');
        }
        else {
            return this.getNameParts().join('_');
        }
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'expression', visitor, options);
        }
    }
}
exports.NamespacedVariableNameExpression = NamespacedVariableNameExpression;
class DottedGetExpression extends Expression {
    constructor(obj, name, dot) {
        super();
        this.obj = obj;
        this.name = name;
        this.dot = dot;
        this.range = util_1.default.createRangeFromPositions(this.obj.range.start, this.name.range.end);
    }
    transpile(state) {
        //if the callee starts with a namespace name, transpile the name
        if (state.file.calleeStartsWithNamespace(this)) {
            return new NamespacedVariableNameExpression(this).transpile(state);
        }
        else {
            return [
                ...this.obj.transpile(state),
                '.',
                new source_map_1.SourceNode(this.name.range.start.line + 1, this.name.range.start.character, state.pathAbsolute, this.name.text)
            ];
        }
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'obj', visitor, options);
        }
    }
}
exports.DottedGetExpression = DottedGetExpression;
class XmlAttributeGetExpression extends Expression {
    constructor(obj, name, at) {
        super();
        this.obj = obj;
        this.name = name;
        this.at = at;
        this.range = util_1.default.createRangeFromPositions(this.obj.range.start, this.name.range.end);
    }
    transpile(state) {
        return [
            ...this.obj.transpile(state),
            '@',
            new source_map_1.SourceNode(this.name.range.start.line + 1, this.name.range.start.character, state.pathAbsolute, this.name.text)
        ];
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'obj', visitor, options);
        }
    }
}
exports.XmlAttributeGetExpression = XmlAttributeGetExpression;
class IndexedGetExpression extends Expression {
    constructor(obj, index, openingSquare, closingSquare) {
        super();
        this.obj = obj;
        this.index = index;
        this.openingSquare = openingSquare;
        this.closingSquare = closingSquare;
        this.range = util_1.default.createRangeFromPositions(this.obj.range.start, this.closingSquare.range.end);
    }
    transpile(state) {
        return [
            ...this.obj.transpile(state),
            new source_map_1.SourceNode(this.openingSquare.range.start.line + 1, this.openingSquare.range.start.character, state.pathAbsolute, '['),
            ...this.index.transpile(state),
            new source_map_1.SourceNode(this.closingSquare.range.start.line + 1, this.closingSquare.range.start.character, state.pathAbsolute, ']')
        ];
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'obj', visitor, options);
            visitors_1.walk(this, 'index', visitor, options);
        }
    }
}
exports.IndexedGetExpression = IndexedGetExpression;
class GroupingExpression extends Expression {
    constructor(tokens, expression) {
        super();
        this.tokens = tokens;
        this.expression = expression;
        this.range = util_1.default.createRangeFromPositions(this.tokens.left.range.start, this.tokens.right.range.end);
    }
    transpile(state) {
        return [
            new source_map_1.SourceNode(this.tokens.left.range.start.line + 1, this.tokens.left.range.start.character, state.pathAbsolute, '('),
            ...this.expression.transpile(state),
            new source_map_1.SourceNode(this.tokens.right.range.start.line + 1, this.tokens.right.range.start.character, state.pathAbsolute, ')')
        ];
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'expression', visitor, options);
        }
    }
}
exports.GroupingExpression = GroupingExpression;
class LiteralExpression extends Expression {
    constructor(token) {
        super();
        this.token = token;
        this.type = util_1.default.tokenToBscType(token);
    }
    get range() {
        return this.token.range;
    }
    transpile(state) {
        let text;
        if (this.token.kind === lexer_1.TokenKind.TemplateStringQuasi) {
            //wrap quasis with quotes (and escape inner quotemarks)
            text = `"${this.token.text.replace(/"/g, '""')}"`;
        }
        else if (reflection_1.isStringType(this.type)) {
            text = this.token.text;
            //add trailing quotemark if it's missing. We will have already generated a diagnostic for this.
            if (text.endsWith('"') === false) {
                text += '"';
            }
        }
        else {
            text = this.token.text;
        }
        return [
            new source_map_1.SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, text)
        ];
    }
    walk(visitor, options) {
        //nothing to walk
    }
}
exports.LiteralExpression = LiteralExpression;
/**
 * This is a special expression only used within template strings. It exists so we can prevent producing lots of empty strings
 * during template string transpile by identifying these expressions explicitly and skipping the bslib_toString around them
 */
class EscapedCharCodeLiteralExpression extends Expression {
    constructor(token) {
        super();
        this.token = token;
        this.range = token.range;
    }
    transpile(state) {
        return [
            new source_map_1.SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, `chr(${this.token.charCode})`)
        ];
    }
    walk(visitor, options) {
        //nothing to walk
    }
}
exports.EscapedCharCodeLiteralExpression = EscapedCharCodeLiteralExpression;
class ArrayLiteralExpression extends Expression {
    constructor(elements, open, close) {
        super();
        this.elements = elements;
        this.open = open;
        this.close = close;
        this.range = util_1.default.createRangeFromPositions(this.open.range.start, this.close.range.end);
    }
    transpile(state) {
        let result = [];
        result.push(new source_map_1.SourceNode(this.open.range.start.line + 1, this.open.range.start.character, state.pathAbsolute, '['));
        let hasChildren = this.elements.length > 0;
        state.blockDepth++;
        for (let i = 0; i < this.elements.length; i++) {
            let previousElement = this.elements[i - 1];
            let element = this.elements[i];
            if (reflection_1.isCommentStatement(element)) {
                //if the comment is on the same line as opening square or previous statement, don't add newline
                if (util_1.default.linesTouch(this.open, element) || util_1.default.linesTouch(previousElement, element)) {
                    result.push(' ');
                }
                else {
                    result.push('\n', state.indent());
                }
                state.lineage.unshift(this);
                result.push(element.transpile(state));
                state.lineage.shift();
            }
            else {
                result.push('\n');
                result.push(state.indent(), ...element.transpile(state));
                //add a comma if we know there will be another non-comment statement after this
                for (let j = i + 1; j < this.elements.length; j++) {
                    let el = this.elements[j];
                    //add a comma if there will be another element after this
                    if (reflection_1.isCommentStatement(el) === false) {
                        result.push(',');
                        break;
                    }
                }
            }
        }
        state.blockDepth--;
        //add a newline between open and close if there are elements
        if (hasChildren) {
            result.push('\n');
            result.push(state.indent());
        }
        result.push(new source_map_1.SourceNode(this.close.range.start.line + 1, this.close.range.start.character, state.pathAbsolute, ']'));
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            for (let i = 0; i < this.elements.length; i++) {
                visitors_1.walk(this.elements, i, visitor, options, this);
            }
        }
    }
}
exports.ArrayLiteralExpression = ArrayLiteralExpression;
class AAMemberExpression extends Expression {
    constructor(keyToken, colonToken, 
    /** The expression evaluated to determine the member's initial value. */
    value) {
        super();
        this.keyToken = keyToken;
        this.colonToken = colonToken;
        this.value = value;
        this.range = util_1.default.createRangeFromPositions(keyToken.range.start, this.value.range.end);
    }
    transpile(state) {
        //TODO move the logic from AALiteralExpression loop into this function
        return [];
    }
    walk(visitor, options) {
        visitors_1.walk(this, 'value', visitor, options);
    }
}
exports.AAMemberExpression = AAMemberExpression;
class AALiteralExpression extends Expression {
    constructor(elements, open, close) {
        super();
        this.elements = elements;
        this.open = open;
        this.close = close;
        this.range = util_1.default.createRangeFromPositions(this.open.range.start, this.close.range.end);
    }
    transpile(state) {
        let result = [];
        //open curly
        result.push(new source_map_1.SourceNode(this.open.range.start.line + 1, this.open.range.start.character, state.pathAbsolute, this.open.text));
        let hasChildren = this.elements.length > 0;
        //add newline if the object has children and the first child isn't a comment starting on the same line as opening curly
        if (hasChildren && (reflection_1.isCommentStatement(this.elements[0]) === false || !util_1.default.linesTouch(this.elements[0], this.open))) {
            result.push('\n');
        }
        state.blockDepth++;
        for (let i = 0; i < this.elements.length; i++) {
            let element = this.elements[i];
            let previousElement = this.elements[i - 1];
            let nextElement = this.elements[i + 1];
            //don't indent if comment is same-line
            if (reflection_1.isCommentStatement(element) &&
                (util_1.default.linesTouch(this.open, element) || util_1.default.linesTouch(previousElement, element))) {
                result.push(' ');
                //indent line
            }
            else {
                result.push(state.indent());
            }
            //render comments
            if (reflection_1.isCommentStatement(element)) {
                result.push(...element.transpile(state));
            }
            else {
                //key
                result.push(new source_map_1.SourceNode(element.keyToken.range.start.line + 1, element.keyToken.range.start.character, state.pathAbsolute, element.keyToken.text));
                //colon
                result.push(new source_map_1.SourceNode(element.colonToken.range.start.line + 1, element.colonToken.range.start.character, state.pathAbsolute, ':'), ' ');
                //determine if comments are the only members left in the array
                let onlyCommentsRemaining = true;
                for (let j = i + 1; j < this.elements.length; j++) {
                    if (reflection_1.isCommentStatement(this.elements[j]) === false) {
                        onlyCommentsRemaining = false;
                        break;
                    }
                }
                //value
                result.push(...element.value.transpile(state));
                //add trailing comma if not final element (excluding comments)
                if (i !== this.elements.length - 1 && onlyCommentsRemaining === false) {
                    result.push(',');
                }
            }
            //if next element is a same-line comment, skip the newline
            if (nextElement && reflection_1.isCommentStatement(nextElement) && nextElement.range.start.line === element.range.start.line) {
                //add a newline between statements
            }
            else {
                result.push('\n');
            }
        }
        state.blockDepth--;
        //only indent the closing curly if we have children
        if (hasChildren) {
            result.push(state.indent());
        }
        //close curly
        result.push(new source_map_1.SourceNode(this.close.range.start.line + 1, this.close.range.start.character, state.pathAbsolute, this.close.text));
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            for (let i = 0; i < this.elements.length; i++) {
                if (reflection_1.isCommentStatement(this.elements[i])) {
                    visitors_1.walk(this.elements, i, visitor, options, this);
                }
                else {
                    visitors_1.walk(this.elements, i, visitor, options, this);
                }
            }
        }
    }
}
exports.AALiteralExpression = AALiteralExpression;
class UnaryExpression extends Expression {
    constructor(operator, right) {
        super();
        this.operator = operator;
        this.right = right;
        this.range = util_1.default.createRangeFromPositions(this.operator.range.start, this.right.range.end);
    }
    transpile(state) {
        return [
            new source_map_1.SourceNode(this.operator.range.start.line + 1, this.operator.range.start.character, state.pathAbsolute, this.operator.text),
            ' ',
            ...this.right.transpile(state)
        ];
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'right', visitor, options);
        }
    }
}
exports.UnaryExpression = UnaryExpression;
class VariableExpression extends Expression {
    constructor(name, namespaceName) {
        super();
        this.name = name;
        this.namespaceName = namespaceName;
        this.range = this.name.range;
    }
    getName(parseMode) {
        return parseMode === Parser_1.ParseMode.BrightScript ? this.name.text : this.name.text;
    }
    transpile(state) {
        var _a;
        let result = [];
        //if the callee is the name of a known namespace function
        if (state.file.calleeIsKnownNamespaceFunction(this, (_a = this.namespaceName) === null || _a === void 0 ? void 0 : _a.getName(Parser_1.ParseMode.BrighterScript))) {
            result.push(new source_map_1.SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, `${this.namespaceName.getName(Parser_1.ParseMode.BrightScript)}_${this.getName(Parser_1.ParseMode.BrightScript)}`));
            //transpile  normally
        }
        else {
            result.push(new source_map_1.SourceNode(this.name.range.start.line + 1, this.name.range.start.character, state.pathAbsolute, this.name.text));
        }
        return result;
    }
    walk(visitor, options) {
        //nothing to walk
    }
}
exports.VariableExpression = VariableExpression;
class SourceLiteralExpression extends Expression {
    constructor(token) {
        super();
        this.token = token;
        this.range = token.range;
    }
    getFunctionName(state, parseMode) {
        let func = state.file.getFunctionScopeAtPosition(this.token.range.start).func;
        let nameParts = [];
        while (func.parentFunction) {
            let index = func.parentFunction.childFunctionExpressions.indexOf(func);
            nameParts.unshift(`anon${index}`);
            func = func.parentFunction;
        }
        //get the index of this function in its parent
        nameParts.unshift(func.functionStatement.getName(parseMode));
        return nameParts.join('$');
    }
    transpile(state) {
        let text;
        switch (this.token.kind) {
            case lexer_1.TokenKind.SourceFilePathLiteral:
                text = `"${fileUrl(state.pathAbsolute)}"`;
                break;
            case lexer_1.TokenKind.SourceLineNumLiteral:
                text = `${this.token.range.start.line + 1}`;
                break;
            case lexer_1.TokenKind.FunctionNameLiteral:
                text = `"${this.getFunctionName(state, Parser_1.ParseMode.BrightScript)}"`;
                break;
            case lexer_1.TokenKind.SourceFunctionNameLiteral:
                text = `"${this.getFunctionName(state, Parser_1.ParseMode.BrighterScript)}"`;
                break;
            case lexer_1.TokenKind.SourceLocationLiteral:
                text = `"${fileUrl(state.pathAbsolute)}:${this.token.range.start.line + 1}"`;
                break;
            case lexer_1.TokenKind.PkgPathLiteral:
                let pkgPath1 = `pkg:/${state.file.pkgPath}`
                    .replace(/\\/g, '/')
                    .replace(/\.bs$/i, '.brs');
                text = `"${pkgPath1}"`;
                break;
            case lexer_1.TokenKind.PkgLocationLiteral:
                let pkgPath2 = `pkg:/${state.file.pkgPath}`
                    .replace(/\\/g, '/')
                    .replace(/\.bs$/i, '.brs');
                text = `"${pkgPath2}:" + str(LINE_NUM)`;
                break;
            case lexer_1.TokenKind.LineNumLiteral:
            default:
                //use the original text (because it looks like a variable)
                text = this.token.text;
                break;
        }
        return [
            new source_map_1.SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, text)
        ];
    }
    walk(visitor, options) {
        //nothing to walk
    }
}
exports.SourceLiteralExpression = SourceLiteralExpression;
/**
 * This expression transpiles and acts exactly like a CallExpression,
 * except we need to uniquely identify these statements so we can
 * do more type checking.
 */
class NewExpression extends Expression {
    constructor(newKeyword, call) {
        super();
        this.newKeyword = newKeyword;
        this.call = call;
        this.range = util_1.default.createRangeFromPositions(this.newKeyword.range.start, this.call.range.end);
    }
    /**
     * The name of the class to initialize (with optional namespace prefixed)
     */
    get className() {
        //the parser guarantees the callee of a new statement's call object will be
        //a NamespacedVariableNameExpression
        return this.call.callee;
    }
    get namespaceName() {
        return this.call.namespaceName;
    }
    transpile(state) {
        var _a;
        const cls = state.file.getClassByName(this.className.getName(Parser_1.ParseMode.BrighterScript), (_a = this.namespaceName) === null || _a === void 0 ? void 0 : _a.getName(Parser_1.ParseMode.BrighterScript));
        //new statements within a namespace block can omit the leading namespace if the class resides in that same namespace.
        //So we need to figure out if this is a namespace-omitted class, or if this class exists without a namespace.
        return this.call.transpile(state, cls === null || cls === void 0 ? void 0 : cls.getName(Parser_1.ParseMode.BrightScript));
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'call', visitor, options);
        }
    }
}
exports.NewExpression = NewExpression;
class CallfuncExpression extends Expression {
    constructor(callee, operator, methodName, openingParen, args, closingParen) {
        var _a, _b, _c;
        super();
        this.callee = callee;
        this.operator = operator;
        this.methodName = methodName;
        this.openingParen = openingParen;
        this.args = args;
        this.closingParen = closingParen;
        this.range = util_1.default.createRangeFromPositions(callee.range.start, ((_c = (_b = (_a = closingParen !== null && closingParen !== void 0 ? closingParen : args[args.length - 1]) !== null && _a !== void 0 ? _a : openingParen) !== null && _b !== void 0 ? _b : methodName) !== null && _c !== void 0 ? _c : operator).range.end);
    }
    transpile(state) {
        let result = [];
        result.push(...this.callee.transpile(state), new source_map_1.SourceNode(this.operator.range.start.line + 1, this.operator.range.start.character, state.pathAbsolute, '.callfunc'), new source_map_1.SourceNode(this.openingParen.range.start.line + 1, this.openingParen.range.start.character, state.pathAbsolute, '('), 
        //the name of the function
        new source_map_1.SourceNode(this.methodName.range.start.line + 1, this.methodName.range.start.character, state.pathAbsolute, `"${this.methodName.text}"`), ', ');
        //transpile args
        //callfunc with zero args never gets called, so pass invalid as the first parameter if there are no args
        if (this.args.length === 0) {
            result.push('invalid');
        }
        else {
            for (let i = 0; i < this.args.length; i++) {
                //add comma between args
                if (i > 0) {
                    result.push(', ');
                }
                let arg = this.args[i];
                result.push(...arg.transpile(state));
            }
        }
        result.push(new source_map_1.SourceNode(this.closingParen.range.start.line + 1, this.closingParen.range.start.character, state.pathAbsolute, ')'));
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            visitors_1.walk(this, 'callee', visitor, options);
            for (let i = 0; i < this.args.length; i++) {
                visitors_1.walk(this.args, i, visitor, options, this);
            }
        }
    }
}
exports.CallfuncExpression = CallfuncExpression;
/**
 * Since template strings can contain newlines, we need to concatenate multiple strings together with chr() calls.
 * This is a single expression that represents the string contatenation of all parts of a single quasi.
 */
class TemplateStringQuasiExpression extends Expression {
    constructor(expressions) {
        super();
        this.expressions = expressions;
        this.range = util_1.default.createRangeFromPositions(this.expressions[0].range.start, this.expressions[this.expressions.length - 1].range.end);
    }
    transpile(state, skipEmptyStrings = true) {
        let result = [];
        let plus = '';
        for (let expression of this.expressions) {
            //skip empty strings
            //TODO what does an empty string literal expression look like?
            if (expression.token.text === '' && skipEmptyStrings === true) {
                continue;
            }
            result.push(plus, ...expression.transpile(state));
            plus = ' + ';
        }
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            for (let i = 0; i < this.expressions.length; i++) {
                visitors_1.walk(this.expressions, i, visitor, options, this);
            }
        }
    }
}
exports.TemplateStringQuasiExpression = TemplateStringQuasiExpression;
class TemplateStringExpression extends Expression {
    constructor(openingBacktick, quasis, expressions, closingBacktick) {
        super();
        this.openingBacktick = openingBacktick;
        this.quasis = quasis;
        this.expressions = expressions;
        this.closingBacktick = closingBacktick;
        this.range = util_1.default.createRangeFromPositions(quasis[0].range.start, quasis[quasis.length - 1].range.end);
    }
    transpile(state) {
        if (this.quasis.length === 1 && this.expressions.length === 0) {
            return this.quasis[0].transpile(state);
        }
        let result = [];
        //wrap the expression in parens to readability
        // result.push(
        //     new SourceNode(
        //         this.openingBacktick.range.start.line + 1,
        //         this.openingBacktick.range.start.character,
        //         state.pathAbsolute,
        //         '('
        //     )
        // );
        let plus = '';
        //helper function to figure out when to include the plus
        function add(...items) {
            if (items.length > 0) {
                result.push(plus, ...items);
            }
            plus = ' + ';
        }
        for (let i = 0; i < this.quasis.length; i++) {
            let quasi = this.quasis[i];
            let expression = this.expressions[i];
            add(...quasi.transpile(state));
            if (expression) {
                //skip the toString wrapper around certain expressions
                if (reflection_1.isEscapedCharCodeLiteralExpression(expression) ||
                    (reflection_1.isLiteralExpression(expression) && reflection_1.isStringType(expression.type))) {
                    add(...expression.transpile(state));
                    //wrap all other expressions with a bslib_toString call to prevent runtime type mismatch errors
                }
                else {
                    add('bslib_toString(', ...expression.transpile(state), ')');
                }
            }
        }
        //wrap the expression in parens to readability
        // result.push(
        //     new SourceNode(
        //         this.openingBacktick.range.end.line + 1,
        //         this.openingBacktick.range.end.character,
        //         state.pathAbsolute,
        //         ')'
        //     )
        // );
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            //walk the quasis and expressions in left-to-right order
            for (let i = 0; i < this.quasis.length; i++) {
                visitors_1.walk(this.quasis, i, visitor, options, this);
                //this skips the final loop iteration since we'll always have one more quasi than expression
                if (this.expressions[i]) {
                    visitors_1.walk(this.expressions, i, visitor, options, this);
                }
            }
        }
    }
}
exports.TemplateStringExpression = TemplateStringExpression;
class TaggedTemplateStringExpression extends Expression {
    constructor(tagName, openingBacktick, quasis, expressions, closingBacktick) {
        super();
        this.tagName = tagName;
        this.openingBacktick = openingBacktick;
        this.quasis = quasis;
        this.expressions = expressions;
        this.closingBacktick = closingBacktick;
        this.range = util_1.default.createRangeFromPositions(quasis[0].range.start, quasis[quasis.length - 1].range.end);
    }
    transpile(state) {
        let result = [];
        result.push(new source_map_1.SourceNode(this.tagName.range.start.line + 1, this.tagName.range.start.character, state.pathAbsolute, this.tagName.text), '([');
        //add quasis as the first array
        for (let i = 0; i < this.quasis.length; i++) {
            let quasi = this.quasis[i];
            //separate items with a comma
            if (i > 0) {
                result.push(', ');
            }
            result.push(...quasi.transpile(state, false));
        }
        result.push('], [');
        //add expressions as the second array
        for (let i = 0; i < this.expressions.length; i++) {
            let expression = this.expressions[i];
            if (i > 0) {
                result.push(', ');
            }
            result.push(...expression.transpile(state));
        }
        result.push(new source_map_1.SourceNode(this.closingBacktick.range.end.line + 1, this.closingBacktick.range.end.character, state.pathAbsolute, '])'));
        return result;
    }
    walk(visitor, options) {
        if (options.walkMode & visitors_1.InternalWalkMode.walkExpressions) {
            //walk the quasis and expressions in left-to-right order
            for (let i = 0; i < this.quasis.length; i++) {
                visitors_1.walk(this.quasis, i, visitor, options, this);
                //this skips the final loop iteration since we'll always have one more quasi than expression
                if (this.expressions[i]) {
                    visitors_1.walk(this.expressions, i, visitor, options, this);
                }
            }
        }
    }
}
exports.TaggedTemplateStringExpression = TaggedTemplateStringExpression;
class AnnotationExpression extends Expression {
    constructor(atToken, nameToken) {
        super();
        this.atToken = atToken;
        this.nameToken = nameToken;
        this.name = nameToken.text;
        this.range = util_1.default.createRangeFromPositions(atToken.range.start, nameToken.range.end);
    }
    /**
     * Convert annotation arguments to JavaScript types
     * @param strict If false, keep Expression objects not corresponding to JS types
     */
    getArguments(strict = true) {
        if (!this.call) {
            return [];
        }
        return this.call.args.map(e => expressionToValue(e, strict));
    }
    transpile(state) {
        return [];
    }
    walk(visitor, options) {
        //nothing to walk
    }
}
exports.AnnotationExpression = AnnotationExpression;
function expressionToValue(expr, strict) {
    if (!expr) {
        return null;
    }
    if (reflection_1.isLiteralString(expr)) {
        //remove leading and trailing quotes
        return expr.token.text.replace(/^"/, '').replace(/"$/, '');
    }
    if (reflection_1.isLiteralNumber(expr)) {
        if (reflection_1.isIntegerType(expr.type) || reflection_1.isLongIntegerType(expr.type)) {
            return parseInt(expr.token.text);
        }
        else {
            return parseFloat(expr.token.text);
        }
    }
    if (reflection_1.isLiteralBoolean(expr)) {
        return expr.token.text.toLowerCase() === 'true';
    }
    if (reflection_1.isArrayLiteralExpression(expr)) {
        return expr.elements
            .filter(e => !reflection_1.isCommentStatement(e))
            .map(e => expressionToValue(e, strict));
    }
    if (reflection_1.isAALiteralExpression(expr)) {
        return expr.elements.reduce((acc, e) => {
            if (!reflection_1.isCommentStatement(e)) {
                acc[e.keyToken.text] = expressionToValue(e.value, strict);
            }
            return acc;
        }, {});
    }
    return strict ? null : expr;
}
